File: src/filter.py
import sys
import matplotlib.pyplot as plt
import numpy as np

import pyanalid

USAGE="""
Filter LIDAR observations

Removes all plots that have fewer than K observations, from all the remaining (populated) plots
removes P percentiles (given in percents, say 5) of observations on both tails of distribution

USAGE:
    python3 ./filter.py input-file K P

EXAMPLE:
    python3 ./filter.py lidar_data.csv.gz 50 5
"""

if len(sys.argv) == 1 or sys.argv[1] == "-h" or sys.argv[1] == "--help":
    print(USAGE)
    sys.exit(0)

lidar_file = sys.argv[1]
min_obs = int(sys.argv[2])
percent = float(sys.argv[3])

pts = pyanalid.read_points(lidar_file)
grid = pyanalid.Grid(5.0, 5.0, pts)

for key in grid.keys():
    n = grid.count_points(key)
    if n < min_obs: continue
    i_from = int(n*percent / 100.0)
    i_to = int(n - i_from)
    pts = grid.points(key)
    for point in pts[i_from:i_to]:
        print(point)

File: src/filter.rs
use clap::{Parser};

use bioshell_statistics::{Histogram, OnlineMultivariateStatistics};

use analid::{Grid, read_points};

#[derive(Parser, Debug)]
#[clap(name = "lidar")]
#[clap(version = "0.2")]
#[clap(about = "Simple analysis of LIDAR measurements", long_about = None)]
struct Args {
    /// staring conformation in the CSV format
    #[clap(short, long, default_value = "", short='f')]
    infile: String,
    /// plot size in meters
    #[clap(short, long, default_value_t = 5.0, short='w')]
    bin_width: f64,
    /// minimum number of observations per plot to print it; plots with fewer measurements will be discarded
    #[clap(short, long, default_value_t = 50, short='m')]
    min_observations: usize,
    /// percentile of observations to trim off; by default 5% observations from each plot are removed from both tails
    #[clap(short, long, default_value_t = 5.0, short='p')]
    percent: f64,
}

fn main() {

    let args = Args::parse();
    let points = read_points(&args.infile);
    let bin_width = args.bin_width;
    let grid: Grid = Grid::new(bin_width, bin_width, points);

    for key in grid.data().keys() {
        let n = grid.count_points(key);
        if n < args.min_observations { continue; }
        let i_from = (n as f64 * args.percent / 100.0) as usize;
        let i_to = (n - i_from)  as usize;

        for point in &grid.points(key).unwrap()[i_from..i_to] {
            println!("{}",point);
        }
    }
}
